### ISSUE TITLE
[Core Profiler] Reintroduce snapshot tests

### ISSUE THREAD

**Issue by rjchow** [woothemes/woocommerce#48248]
**Issue Description:**

In the [Core Profiler XState V5 migration PR](https://github.com/woocommerce/woocommerce/pull/48135), we removed the snapshot tests as @xstate/test was deprecated and most of its functionality was moved into @xstate/graph. However the functionality that we relied on was generating test plans, which was removed.

It is possible to compose the test plans using @xstate/graph functionality, however the documentation has not been updated for the @xstate/graph@2.0.0 release. Lets reintroduce the snapshot tests when the documentation is ready (~approx within the next month).

**Comment by adrianduffell** [comment]
> Lets reintroduce the snapshot tests when the documentation is ready (~approx within the next month).
I'll just put this on hold until we have the documentation ready,

**Comment by adrianduffell** [comment]
Ha not sure what my previous comment was about.

@rjchow, is the documentation now ready?

**Comment by rjchow** [comment]
No 😔 Seems like they’ve had layoffs and it’s mostly just davidkpiano (founder) making contributions now

**Comment by timmy5685** [comment]
Hi @rjchow - thanks for such a detailed report, it's really helpful.

I know this has been open for a long time, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

I'd like to get an engineer's eyes on this before we make a call, so I've flagged it for developer feedback. Someone will take a proper look at the repro steps and how it's behaving now.

The bit that needs a closer look: The missing piece is an engineering decision about how to replace the removed test-plan functionality, not a product call.

**Comment by vladolaru** [comment]
I checked this against current trunk and the current XState tooling. This is still valid, but the original reason for keeping it on hold looks stale.

The coverage removed in #48135 has not been restored. #38154 added the old `plugins/woocommerce-admin/client/core-profiler/test/core-profiler-machine.test.tsx` model test plus the large `core-profiler-machine.test.tsx.snap` file; #48135 deleted both during the XState v5 migration. On trunk, the Core Profiler still exports an XState v5 machine via [`coreProfilerStateMachineDefinition = createMachine( ... )`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/admin/client/core-profiler/index.tsx#L709-L760), and `CoreProfilerController` still provides action/actor override seams around that machine ([controller setup](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/admin/client/core-profiler/index.tsx#L1778-L1796)). Current tests cover pages, services, navigation, plugin UI, and Tracks actions, but I did not find restored machine/path coverage under the current Core Profiler tree.

The XState situation has changed since this was opened. Current XState graph docs say graph utilities are part of the main `xstate` package and should be imported from [`xstate/graph`](https://stately.ai/docs/graph), rather than relying on the deprecated `@xstate/test` package. In this repo specifically, the admin package still keeps XState v4 as `xstate` and aliases XState v5 as [`xstate5`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/admin/package.json#L124-L128), so the practical candidate here is `xstate5/graph`, not `xstate/graph`. The admin package still carries deprecated [`@xstate/test@0.5.1`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/admin/package.json#L173-L174).

I would keep this open, remove `status: on hold`, and re-scope the work from "reintroduce snapshot tests" to "restore XState v5-compatible Core Profiler machine/path coverage."

The implementation path I would try:

- Start with a small proof of concept using `createTestModel` from `xstate5/graph`.
- Prefer shortest-path coverage first, so we prove reachability without generating an unmaintainable number of UI-driven paths.
- Use simple-path coverage only for a deliberately bounded subset if transition coverage is needed.
- Structure the tests around `path.test( { events, states } )`: event handlers drive the relevant UI/action/actor transition, and state handlers assert reached page/state markers plus critical side effects.
- Do not restore the broad full-page snapshot file by default. Use targeted DOM, state, mocked actor/action, and side-effect assertions; add small snapshots only where they protect a specific regression.
- Keep #48416 separate for Tracks instrumentation. The same path scaffold may help that issue later, but #48248 should first restore machine/path coverage.
- Remove `@xstate/test` once the replacement lands if it remains unused.

So the issue still tracks real test debt, but the next step is an implementation decision using current XState v5 graph tooling, not waiting for upstream docs or recreating the old snapshot suite as-is.
